jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannelOverHttp.java

```diff
@@ -147,12 +241,94 @@ public void earlyEOF()
         _httpConnection.getGenerator().setPersistent(false);
         // If we have no request yet, just close
         if (_metadata == null)
+        {
             _httpConnection.close();
-        else if (onEarlyEOF() || _delayedForContent)
+        }
+        else
         {
-            _delayedForContent = false;
-            handle();
+            markEarlyEOF();
+            if (_delayedForContent)
+            {
+                _delayedForContent = false;
+                handle();
+            }
+        }
+    }
+
+    private void markEarlyEOF()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("received early EOF, content = {}", _content);
+        EofException failure = new EofException("Early EOF");
+        if (_content != null)
+            _content.failed(failure);
+        _content = new HttpInput.ErrorContent(failure);
+    }
+
+    @Override
+    protected boolean eof()
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("received EOF, content = {}", _content);
+        if (_content == null)
+        {
+            _content = EOF;
+        }
+        else
+        {
+            HttpInput.Content c = _content;
+            _content = new HttpInput.Content(c.getByteBuffer())
```

Comment:
Both HTTP/2 and HTTP/1.1 wrap an existing content, so it's worth having a proper wrapper class with the ability to specify `eof`, rather than anonymous classes -- they will also print better when DEBUG is enabled.

Reply:
Done.